### ISSUE TITLE
Changing the compiler to ghc784 help needed

### ISSUE THREAD

**Issue by 573** [Gabriel439/haskell-nix#69]
Hi  @Gabriel439,

using the apporach described [here](https://github.com/Gabriel439/haskell-nix/tree/master/project1#changing-the-compiler) I tried to build [flickr haskell](https://github.com/sof/flickr.git).

Basically modifying to use the correct network-uri and mime versions works, but now I am stuck at the ghc-Version to use, because the flickr binding is using a now incompatible (["Make Applicative a superclass of Monad"](https://hackage.haskell.org/package/base-4.12.0.0/changelog)) implementation approach.

My hope was that specifying `--argstr compiler ghc784` solved it but that basically lead me nowhere:

> error: attribute 'ghc784' missing, at /home/vagrant/flickr/release5.nix:8:27

(the mentioned release5.nix being bascially the same as in 384e1d6 )

Seems a bit similar to #15, but I can't see it through.

**Comment by Gabriel439** [comment]
The set of available GHC versions that `nixpkgs` provides changes depending on the revision.  If you know what revision of `nixpkgs` you are building against you can check the set of available GHC compilers by visiting this page:

```
https://github.com/NixOS/nixpkgs/blob/${REVISION}/pkgs/top-level/haskell-packages.nix
```

For example, currently the available GHC versions listed at https://github.com/NixOS/nixpkgs/blob/70178169fa562c85a7d6b93e7b6997a85c7f7c7f/pkgs/top-level/haskell-packages.nix#L47-L67 are `ghc822`, `ghc844`, `ghc861`, and `ghc862`

**Comment by 573** [comment]
Thanks for quick responding, Gabriel.  

I think I got my head around it a little closer now:  

> [vagrant@nixbox:\~]$ sudo nix-channel --add https://github.com/NixOS/nixpkgs/archive/release-18.03.tar.gz oldghcs  
[vagrant@nixbox:\~]$ sudo nix-channel --update oldghcs  
[vagrant@nixbox:\~]$ nix-shell '\<oldghcs\>' -p ghc  # this does not bring ghc-7.8.4 yet, although it appears in the following expressions output. had to investigate further     
[nix-shell:\~]$ nix-instantiate --eval -E "with import \<nixpkgs\> {}; lib.attrNames haskell.compiler"  
[ "ghc6102Binary" "ghc6104" "ghc6123" "ghc704" "ghc704Binary" "ghc7103" "ghc7103Binary" "ghc722" "ghc742" "ghc742Binary" "ghc763" "ghc784" "ghc784Binary" "ghc802" "ghc821Binary" "ghc822" "ghc841" "ghcHEAD" "ghcHaLVM240" "ghcjs" "ghcjsHEAD" "integer-simple" "uhc" ]  

Right then I found:  

> ~[vagrant@nixbox:\~]$ NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/release-18.03.tar.gz nix-shell -p haskell.compiler.ghc784  #~ but it even works w/o the NIX_PATH prefix  
[vagrant@nixbox:\~]$ nix-shell '\<oldghcs\>' -p haskell.compiler.ghc784 # and yes this works even w/o the 'oldghcs' tag  
[nix-shell:\~]$ ghc --version  
The Glorious Glasgow Haskell Compilation System, version 7.8.4  
[nix-shell:\~]$ nix-instantiate --eval -E "with import \<oldghcs\> {}; lib.attrNames haskell.compiler"
[ "ghc704" "ghc704Binary" "ghc7103" "ghc7103Binary" "ghc742" "ghc742Binary" "ghc763" "ghc784" "ghc784Binary" "ghc802" "ghc821Binary" "ghc822" "ghc841" "ghc843" "ghcHEAD" "ghcjs" "ghcjsHEAD" "integer-simple" ]  

Long story short: Ok for me. Now have to try my use case above again.

**Comment by 573** [comment]
Changed one line in release5.nix to contain `pkgs = import (fetchTarball
https://github.com/NixOS/nixpkgs/archive/release-18.03.tar.gz) { inherit
config; };` now and getting:

> [nix-shell:~/flickr]$ nix-build --argstr compiler ghc784 --attr flickr release5.nix  
error: attribute 'ghc784' missing, at /home/vagrant/flickr/release5.nix:8:27

Also tried with the same effect to change the release5.nix to contain in the header `{ pkgs ? import (fetchTarball
https://github.com/NixOS/nixpkgs/archive/release-18.03.tar.gz) {}
, compiler ? "ghc802" }:`

**Comment by 573** [comment]
Attachments seem not to work on this computer...
network-2.5.0.0.nix:
<pre>{ mkDerivation, base, bytestring, HUnit, parsec, stdenv
, test-framework, test-framework-hunit, test-framework-quickcheck2
, unix
}:
mkDerivation {
  pname = "network";
  version = "2.5.0.0";
  sha256 = "b2166f54af2aed140fbd2c1d7915f8e11eef37b9e33c2cd71618bfe87e7320f5";
  libraryHaskellDepends = [ base bytestring parsec unix ];
  testHaskellDepends = [
    base bytestring HUnit test-framework test-framework-hunit
    test-framework-quickcheck2
  ];
  homepage = "https://github.com/haskell/network";
  description = "Low-level networking interface";
  license = stdenv.lib.licenses.bsd3;
}
</pre>

mime-0.3.4.nix:
<pre>{ mkDerivation, base, stdenv }:
mkDerivation {
  pname = "mime";
  version = "0.3.4";
  sha256 = "9e4b77e95df4eeb7d247acf46c996785a974f248cfa8729300dbf816c5f09bce";
  libraryHaskellDepends = [ base ];
  homepage = "https://github.com/GaloisInc/mime";
  description = "Working with MIME types";
  license = stdenv.lib.licenses.bsd3;
}
</pre>

default.nix:
<pre>{ mkDerivation, base, filepath, HTTP, mime, network, random, stdenv
, utf8-string, xhtml, xml
}:
mkDerivation {
  pname = "flickr";
  version = "0.3.3";
  src = ./.;
  isLibrary = true;
  isExecutable = true;
  libraryHaskellDepends = [
    base filepath HTTP mime network random utf8-string xml
  ];
  executableHaskellDepends = [ xhtml ];
  homepage = "https://github.com/sof/flickr";
  description = "Haskell binding to the Flickr API";
  license = stdenv.lib.licenses.bsd3;
}
</pre>

release.nix:
<pre>{ compiler ? "ghc802" }:

let
  config = {
    packageOverrides = pkgs: rec {
      haskell = pkgs.haskell // {
        packages = pkgs.haskell.packages // {
          "${compiler}" = pkgs.haskell.packages."${compiler}".override {
            overrides = haskellPackagesNew: haskellPackagesOld: rec {
              network =
                haskellPackagesNew.callPackage ./network-2.5.0.0.nix {
};

              flickr =
                haskellPackagesNew.callPackage ./default.nix { };

              mime =
                haskellPackagesNew.callPackage ./mime-0.3.4.nix { };
            };
          };
        };
      };
    };
  };

  pkgs = import <nixpkgs> { inherit config; };

in
  { flickr = pkgs.haskell.packages.${compiler}.flickr;
}
</pre>

**Comment by Gabriel439** [comment]
My mistake, I pointed you in the wrong direction.  The actual place to check is the `packages` field (in the same file):

https://github.com/NixOS/nixpkgs/blob/release-18.03/pkgs/top-level/haskell-packages.nix#L112-L148

... since you are overriding `pkgs.haskell.packages."${compiler}"` and not `pkgs.haskell.compiler."${compiler}"`.  In this case there is no `pkgs.haskell.packages.ghc784`, which is why you got that failure

**Comment by 573** [comment]
Wow, that was a quick 100! Only line I needed to change in my `release.nix` was the one before the `in` clause at the bottom, former which let me process `nix-build --argstr compiler ghc784 --attr flickr release.nix` right on:

> pkgs = import (fetchTarball https://github.com/NixOS/nixpkgs/archive/release-17.03.tar.gz) { inherit config; };

Will have to redo this on a faster computer though to double-check that no nix-channel modification is either needed nor the header of `release.nix` needs to read `{ pkgs ? import (fetchTarball https://github.com/NixOS/nixpkgs/archive/release-17.03.tar.gz) {}, compiler ? "ghc802" }:`. Actually I have both of the latter modifications currently on.

**Comment by 573** [comment]
It works all well without nix-channeling and without mods to the release.nix header as I dared assuming in my last comment.
Again thanks!

**Comment by Gabriel439** [comment]
@573: You're welcome! :slightly_smiling_face:
